### ISSUE TITLE
How to use it in a non-hock environment

### ISSUE THREAD

**Issue by shenhaoguangDayDayUp** [pmndrs/jotai#753]
![image](https://user-images.githubusercontent.com/17617821/135702175-db5036b1-b7f4-46d8-8fcd-f70ef93bfd55.png)
![image](https://user-images.githubusercontent.com/17617821/135702198-e7cdf805-a81d-4d2e-ba98-6bb0254af366.png)

**Comment by dai-shi** [comment]
jotai atom works like `useState`. How would you it with `useState`?

**Comment by shenhaoguangDayDayUp** [comment]
> jotai atom/useAtom works like `useState`. How would you it with `useState`?

Yes, there is no way, but how do I achieve the effect I want, this place emits, and another place monitors the change of this value

**Comment by dai-shi** [comment]
I'm not sure about the debugger in the screenshot, but it has to be global?
useState (as well as useAtom) is not for global variable. So, essentially if you need to read the state value outside React, you would need explicitly bind the state to a global variable.
```js
let globalVariable
const Component = () => {
  const [state, setState] = useState(...)
  useEffect(() => {
    globalVariable = state
  })
  ...
}
```

Jotai has some integrations to extend this pattern: `jotai/zustand`, `jotai/valtio` and `jotai/redux`.

I don't think these are answers your are looking for, though.

**Comment by shenhaoguangDayDayUp** [comment]
Thank you for your answers, I want like this, for example, 
![image](https://user-images.githubusercontent.com/17617821/135735640-e4381c48-ba0c-425c-8a20-d75e4d995da8.png)
and in component  listening value change
![image](https://user-images.githubusercontent.com/17617821/135735658-91ff8bb8-b4fe-4a4f-9f46-1e022783a702.png)

**Comment by shenhaoguangDayDayUp** [comment]
> I'm not sure about the debugger in the screenshot, but it has to be global? useState (as well as useAtom) is not for global variable. So, essentially if you need to read the state value outside React, you would need explicitly bind the state to a global variable.
> 
> ```js
> let globalVariable
> const Component = () => {
>   const [state, setState] = useState(...)
>   useEffect(() => {
>     globalVariable = state
>   })
>   ...
> }
> ```
> 
> Jotai has some integrations to extend this pattern: `jotai/zustand`, `jotai/valtio` and `jotai/redux`.
> 
> I don't think these are answers your are looking for, though.

Thank you for your answers, I want like this, for example, 
![image](https://user-images.githubusercontent.com/17617821/135735698-fbcc85aa-ca73-427b-ad16-03cd7fc4bf47.png)

and in component  listening value change
![image](https://user-images.githubusercontent.com/17617821/135735725-a8ad9d59-2af9-4d55-8e03-83d175d9a0bb.png)

**Comment by dai-shi** [comment]
Okay, as I said it's not possible to do something not possible with useState.

As a workaround, here's a pattern for custom callback on write. (code snippet from https://github.com/pmndrs/jotai/issues/750#issuecomment-932118730)
```js
const baseCountAtom = atom(0) // internal atom not for export
const effectsForCountAtom = atom([]) // list of effect functions
const countAtom = atom(
  (get) => get(baseCountAtom),
  (get, set, arg) => {
    set(baseCountAtom, arg)
    get(effectsForCountAtom).forEach((fn) => fn(get, set)) // can be delayed with setTimeout
  }
)
```

And then, something like this:
```js
  useEffect(() => {
    const callback = (get, set) => {
      console.log('count is changed:', get(countAtom))
    }
    setEffectsForCount(prev => [...prev, callback])
    return () => {
      setEffectsForCount(prev => prev.filter(fn => fn !== callback))
    }
  }, [...])
```

**Comment by shenhaoguangDayDayUp** [comment]
> Okay, as I said it's not possible to do something not possible with useState.
> 
> As a workaround, here's a pattern for custom callback on write. (code snippet from [#750 (comment)](https://github.com/pmndrs/jotai/issues/750#issuecomment-932118730))
> 
> ```js
> const baseCountAtom = atom(0) // internal atom not for export
> const effectsForCountAtom = atom([]) // list of effect functions
> const countAtom = atom(
>   (get) => get(baseCountAtom),
>   (get, set, arg) => {
>     set(baseCountAtom, arg)
>     get(effectsForCountAtom).forEach((fn) => fn(get, set)) // can be delayed with setTimeout
>   }
> )
> ```
> 
> And then, something like this:
> 
> ```js
>   useEffect(() => {
>     const callback = (get, set) => {
>       console.log('count is changed:', get(countAtom))
>     }
>     setEffectsForCount(prev => [...prev, callback])
>     return () => {
>       setEffectsForCount(prev => prev.filter(fn => fn !== callback))
>     }
>   }, [...])
> ```

Hi,I don’t think you understand my question. In fact, I want to use a method (similar to EventBus), or use jotai, to realize communication and value transfer, "emit" is used in the class, and "on" is used in the component

![image](https://user-images.githubusercontent.com/17617821/136486115-8e2242bb-a81d-4746-9d05-3d1abdc118a3.png)
![image](https://user-images.githubusercontent.com/17617821/136486224-4a7f9fba-b0c2-4aa8-9101-982f9bea7e94.png)
Looking forward to your reply,thank you

**Comment by dai-shi** [comment]
Jotai atom values only exist in React tree, so you need `useEffect` to communicate with outside React.
Please make a minimal codesandbox to illustrate what you would like to do with `useState` without jotai, and we can discuss how that can be converted with jotai.

**Comment by shenhaoguangDayDayUp** [comment]
ok .This is demo https://codesandbox.io/s/cocky-lalande-sbx1x

**Comment by shenhaoguangDayDayUp** [comment]
> Jotai atom values only exist in React tree, so you need `useEffect` to communicate with outside React. Please make a minimal codesandbox to illustrate what you would like to do with `useState` without jotai, and we can discuss how that can be converted with jotai.
ok .This is demo https://codesandbox.io/s/cocky-lalande-sbx1x

**Comment by shenhaoguangDayDayUp** [comment]
> > ok .This is demo https://codesandbox.io/s/cocky-lalande-sbx1x
> 
> It was not with `useState`, but your example shows something. Here's something based on it: https://codesandbox.io/s/crazy-albattani-bv9r6?file=/src/App.js (it would need an event queue to allow receiving past events) If this doesn't capture your goal, please create another demo.
Thank you , This is possible,But Can jotai add the scenario I said?

**Comment by dai-shi** [comment]
> Can jotai add the scenario I said?

You mean the scenario you can't implement with useState? No, it's not possible at the moment. We seek some possibilities in the future, but not sure if it will become to cover something you expect.
